lib/FunctionIsolation/IsolateFunctions.cpp

```diff
@@ -207,25 +213,51 @@ class IsolateFunctionsImpl {
   void populateFunctionDispatcher();
 
   /// Create code to throw of an exception
-  void throwException(IRBuilder<> &Builder, StringRef Reason);
+  void throwException(IRBuilder<> &Builder,
+                      StringRef Reason,
+                      const Instruction *OldInstWithDbgLoc = nullptr);
 
-  void throwException(BasicBlock *BB, StringRef Reason) {
+  void throwException(BasicBlock *BB,
+                      StringRef Reason,
+                      const Instruction *OldInstWithDbgLoc = nullptr) {
     IRBuilder<> Builder(BB);
-    throwException(Builder, Reason);
+    throwException(Builder, Reason, OldInstWithDbgLoc);
   }
 };
 
-void IFI::throwException(IRBuilder<> &Builder, StringRef Reason) {
+void IFI::throwException(IRBuilder<> &Builder,
+                         StringRef Reason,
+                         const Instruction *OldInstWithDbgLoc) {
   revng_assert(RaiseException != nullptr);
 
+  BasicBlock *InsertBlock = Builder.GetInsertBlock();
+
+  // Compute the debug location of the newly inserted call.
+  // Give precedence to OldInstWithDbgLoc, and fallback to the Builder
+  // insertion point.
+  DebugLoc DbgLoc;
+  if (OldInstWithDbgLoc) {
+    DbgLoc = OldInstWithDbgLoc->getDebugLoc();
+  } else {
+
+    revng_assert(not InsertBlock->empty());
+    BasicBlock::iterator InsertPoint = Builder.GetInsertPoint();
+    if (InsertPoint != InsertBlock->end())
+      DbgLoc = InsertPoint->getDebugLoc();
+    else
+      DbgLoc = InsertBlock->back().getDebugLoc();
+  }
+  revng_assert(DbgLoc);
```

Comment:
This code is very similar to the one in `createFunctionCall`. Can you factor it out in a function?